cdap-api/src/main/java/io/cdap/cdap/api/app/ApplicationUpdateContext.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright © 2020 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.api.app;
+
+import com.google.gson.JsonSyntaxException;
+import io.cdap.cdap.api.Config;
+import io.cdap.cdap.api.artifact.ArtifactId;
+
+import io.cdap.cdap.api.artifact.ArtifactScope;
+import io.cdap.cdap.api.artifact.ArtifactVersionRange;
+import java.lang.reflect.Type;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * Context for updating Application configs. Provides helper methods for application to support operations like config
+ * upgrade.
+ */
+public interface ApplicationUpdateContext {
+
+  /**
+   * @return All update actions application should perform on the config.
+   */
+  List<ApplicationConfigUpdateAction> getConfigUpdateActions();
```

Comment:
Don't think something like this is needed right now, can leave it up to the app what it wants to do.

Would prefer not to add anything until we have concrete use cases. It's much harder to remove something from the API than to add something.

Reply:
This is added because, we now only have one API to do everything related to updating an api (includes upgrade may be downgrade in future too). This one API will take in list of actions to perform and follow them step by step.

Reply:
I see, because it's now update and not upgrade. ok that's fine.